diff --git a/init/Kconfig b/init/Kconfig
index 47a71fbea..dc89a13de 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1237,7 +1237,6 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 
 config CC_OPTIMIZE_FOR_PERFORMANCE_O3
 	bool "Optimize more for performance (-O3)"
-	depends on ARC
 	imply CC_DISABLE_WARN_MAYBE_UNINITIALIZED  # avoid false positives
 	help
 	  Choosing this option will pass "-O3" to your compiler to optimize
