--- a/glamor/glamor_egl.c	2020-01-09 16:41:24.446281543 +0100
+++ b/glamor/glamor_egl.c	2020-01-09 16:44:15.675156381 +0100
@@ -1007,12 +1007,10 @@
             goto error;
         }
 
-        if (epoxy_gl_version() < 21) {
-            xf86DrvMsg(scrn->scrnIndex, X_INFO,
-                       "glamor: Ignoring GL < 2.1, falling back to GLES.\n");
-            eglDestroyContext(glamor_egl->display, glamor_egl->context);
-            glamor_egl->context = EGL_NO_CONTEXT;
-        }
+        xf86DrvMsg(scrn->scrnIndex, X_INFO,
+                   "glamor: Ignoring GL < 2.1, falling back to GLES.\n");
+        eglDestroyContext(glamor_egl->display, glamor_egl->context);
+        glamor_egl->context = EGL_NO_CONTEXT;
     }
 
     if (glamor_egl->context == EGL_NO_CONTEXT) {
